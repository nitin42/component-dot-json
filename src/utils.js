import { createElement } from 'react'
import ReactHtmlParser from 'react-html-parser'
import isHtml from 'is-html'

/**

TODO:
  * Gracefully handle the errors
  * handlers
  * State

*/

let key = 0

// Get the required data from the JSON schema
const getData = (schema) => ({
  name: schema.name,
  type: schema.type,
  props: schema.props,
})

/**
 * Render the html elements
 */
const createFromHtml = (children) => {
  // This helps us avoid dangerouslySetInnerHtml
  const { type, props } = ReactHtmlParser(children)[0]

  key++
  return createElement(type, { key, ...props }, recurseChildren(props.children))
}

/**
 * Recursively render the children
 */
const recurseChildren = (child) => {
  if (isHtml(child)) {
    return createFromHtml(child)
  }
  
  return child
}

const traverse = (child) => recurseChildren(child)

/**
 * @param {*} children Children array
 */
const renderChildren = (children) => {
  return children.map(child => {
    if (typeof child === 'object') {
      // Use recursion to resolve the children again
      return creator(child)
    }
    
    return traverse(child)
  })
}

/**
 * @param {*} where Component name
 */
const throwError = (where) => {
  throw new Error(
    `property 'children' should be an array in component '${where}'.`
  )
}

/**
 * Resolves children array (if any) of the root element
 * @param {*} msg Corresponds to the component name, used in error reporting.
 * @param {*} props Component props
 */
const resolveChildren = (msg, props) => {
  return props.children && Array.isArray(props.children)
  ? renderChildren(props.children)
  : props.children && !Array.isArray(props.children) ? throwError(msg) : null
}

/**
 * Create React element from JSON
 * @param {*} name Component name
 * @param {*} type Element type
 * @param {*} props Component props
 */
const createReactElement = (name, type, props) => {
  key++

  return createElement(
    type,
    { ...props, key },
    resolveChildren(name, props)
  )
}

export const creator = (schema) => {
  // Root element name, type and props
  const { name, type, props } = getData(schema)

  return createReactElement(name, type, props)
}
