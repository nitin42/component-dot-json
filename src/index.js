import { creator } from './utils'

// Top level function to create the component
function createComponent(schema, components) {
  const ComponentFromJson = () => creator(schema, components)

  // For debugging in react-devtools
  Object.defineProperty(ComponentFromJson, 'name', {
    value: schema.name,
    writable: true
  })

  return ComponentFromJson
}

export { createComponent }
