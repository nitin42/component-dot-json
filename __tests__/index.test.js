import React from 'react'
import renderer from 'react-test-renderer'

import { createComponent } from '../src'

import root from './samples/root.json'
import html from './samples/html.json'
import nested from './samples/nested.json'
import tree from './samples/Tree.json'
import multiple from './samples/multiple.json'
import whole from './samples/whole.json'
import customComp from './samples/customComponent.json'

const Header = (props) => <div>Header {props.children}</div>
const components = {
  Header
}

const RootComp = createComponent(root)
const HtmlComp = createComponent(html)
const NestedComp = createComponent(nested)
const CompTree = createComponent(tree)
const MultComp = createComponent(multiple)
const CustomComp = createComponent(customComp, components)
const WholeComp = createComponent(whole, components)

describe('Component-dot-json', () => {
  it('should parse the root element', () => {
    const tree = renderer.create(<RootComp />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render the html elements', () => {
    const tree = renderer.create(<HtmlComp />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render the nested html elements', () => {
    const tree = renderer.create(<NestedComp />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render a component as a child', () => {
    const tree = renderer.create(<CompTree />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render multiple components as children', () => {
    const tree = renderer.create(<MultComp />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render components, strings and html elements', () => {
    const tree = renderer.create(<WholeComp />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render custom react component', () => {
    const tree = renderer.create(<CustomComp />).toJSON();

    expect(tree).toMatchSnapshot()
  })
})
