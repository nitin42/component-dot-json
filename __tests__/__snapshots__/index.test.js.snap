// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component-dot-json should parse the root element 1`] = `
<div
  style={
    Object {
      "color": "red",
      "fontSize": "12px",
    }
  }
>
  Hello World
</div>
`;

exports[`Component-dot-json should render a component as a child 1`] = `
<div>
  <p>
    Hello World
  </p>
</div>
`;

exports[`Component-dot-json should render components, strings and html elements 1`] = `
<div>
  <div>
    <div>
      Hello
    </div>
  </div>
  Hello World
  <button>
    Click
  </button>
  <div>
    <p>
      Hello
    </p>
  </div>
  <div>
    <div>
      Hello
    </div>
  </div>
  <div>
    Header 
    <div>
      Hello
    </div>
  </div>
</div>
`;

exports[`Component-dot-json should render custom react component 1`] = `
<div>
  Header 
  <h1
    style={
      Object {
        "color": "red",
        "fontSize": "10px",
      }
    }
  >
    Hello World
  </h1>
</div>
`;

exports[`Component-dot-json should render multiple components as children 1`] = `
<div>
  <div>
    <div>
      Hello
    </div>
  </div>
  <div>
    <div>
      Hello
    </div>
  </div>
  <div>
    <div>
      Hello
    </div>
  </div>
</div>
`;

exports[`Component-dot-json should render the html elements 1`] = `
<div>
  <h1
    style={
      Object {
        "color": "red",
        "fontSize": "10px",
      }
    }
  >
    Hello World
  </h1>
</div>
`;

exports[`Component-dot-json should render the nested html elements 1`] = `
<div>
  <h1
    style={
      Object {
        "color": "red",
        "fontSize": "10px",
      }
    }
  >
    <p
      style={
        Object {
          "color": "mistyrose",
          "fontSize": "20px",
        }
      }
    >
      Hello World
    </p>
  </h1>
</div>
`;
